components/watcher/src/watcher.rs

```diff
@@ -141,7 +145,25 @@ impl WatcherData {
                     });
                 }
 
-                let response = process_signed_query(qry, &self.event_storage).unwrap();
+                let result = process_signed_query(qry.clone(), &self.event_storage);
+                let response = match result {
+                    Err(SignedQueryError::QueryError(QueryError::UnknownId { .. })) => {
+                        // forward query to witness
+                        let response = self.forward_query(&qry).await?;
```

Comment:
Events that we got from witness were validated and saved in `forward_message`, but this function returns all of them, without any checks. So event with the wrong signature may be added to the mailbox.

Reply:
`forward_query` now adds replies to mailbox after processing them and ignores invalid ones.